### ISSUE TITLE
Markdown Rendering

### ISSUE THREAD

**Issue by tylerkern** [instructure/canvas-lms#503]
Since getting involved in open source, I've come to love Markdown, and I'm sure I'm not alone. I think it would be awesome if canvas could render Markdown!

**Comment by jfraboni** [comment]
YES PLZ.  The HTML Editor was a nice feature, 10 years ago... Markdown is easy to learn, and super fast.  Plus, it plays nice with other platforms, like GitHub.  What we'd really like to see is the ability to point to a .md file on GitHub, and voila, our lesson plan is magically rendered in Canvas.  Doing so also makes content completely independent from Canvas, which it should be.

**Comment by togakangaroo** [comment]
What needs to happen for this to be prioritized? Are there sprints and sprint planning? Can anyone just code it and submit a PR?

markdown-js is perfectly fine and runs just fine in the browser (and where it doesn't, plain text markdown doesn't look so bad).

If I was to implement it is there anything that I should be aware of? I see the Quick Start says you don't support Windows. If I'm going to set up a VM to develop Canvas I want to make sure a PR would have a reasonably good chance of success.

**Comment by togakangaroo** [comment]
Also of note, Tiny MCE (the rich text editor used here) [actually supports Markdown in their latest version!](http://www.tinymce.com/develop/changelog/?ctrl=version&act=view&pr_id=1&vr_id=869)

Last update to the TinyMCE directory in source was two years ago. So what's the plan?

**Comment by claydiffrient** [comment]
@togakangaroo We have an update to TinyMCE in master right now (b257d4c8c7e39be8d538fdab7cdd94b4a2363a3a).  I am however, unsure if the markdown plugin is enabled. 

cc: @evizitei

**Comment by evizitei** [comment]
the markdown plugin is not enabled out of the box, the first *big* step was getting the upgrade in play, we want to let that bake for a time before we monkey with any other variants to make sure that we've killed absolutely every problem we can related to the upgrade first (assuming that we find any)

**Comment by togakangaroo** [comment]
I see, would it be possible to enable the plugin on our personal instance of canvas? Would this require digging around in the ruby code or just adding the js plugin and configuring?

**Comment by evizitei** [comment]
do you guys run canvas on your own?  If so then, yes you'd want to look in 

https://github.com/instructure/canvas-lms/blob/master/public/javascripts/tinymce.config.js#L52

that's where we specify which plugins we want to use, you'd probably also want to get an AMD wrapper around that plugin for the front-end build (see https://github.com/instructure/canvas-lms/blob/master/public/javascripts/bower/tinymce/plugins/autolink/plugin.js#L12 for an example)  and update https://github.com/instructure/canvas-lms/blob/master/app/coffeescripts/editor/stocktiny.coffee to specify that you want to load that plugin into the build, and to tell tiny not to try to load it on it's own later.

**Comment by pyrogerg** [comment]
Has this feature been implemented? Can I use it?

**Comment by 10by10pixel** [comment]
@pyrogerg No. But it's possible to enable on your own. Look at some of the other comments.

**Comment by elec3647** [comment]
This would be so useful!

**Comment by AlinaWithAFace** [comment]
I feel like markdown support would be great! The html text editor feels like it's a bit overkill

**Comment by rrnewton** [comment]
Arrived here with the same thought -- it would be great to keep content external to Canvas and not copy/paste html between semesters.  A natural format for that content would be markdown, since we use it for everything else (Github pages, issues, etc).

**Comment by schirner** [comment]
Would be very useful indeed!

Just to cross link, there is an [idea](https://community.canvaslms.com/ideas/15037-markdown-support-for-canvas) tracker on Canvas about Markdown rendering.

**Comment by RailRunner166** [comment]
I suggested this a while back on their feedback page but I don't know where it went from there, probably got deleted after a year of inactivity, still would be nice to see though!

**Comment by 0az** [comment]
Having just completed an exam in Canvas, and with the expectation of completing *more* using TinyMCE, I would like to suggest another possible implementation: store raw markdown in `<script type="text/markdown;variant=commonmark">` or `variant=gfm`, depending on if the TinyMCE default of `gfm` or the IMHO superior Commonmark gets used.

This has the advantage of not requiring roundtripping through HTML, which may lead to poor UX. Downside is that rendering to HTML is possibly forced onto the client, though I'm given to believe that it'd be necessary anyway.

However, I'm not familiar with the consumer side, and I don't know if they'd be able to support this without modification. Thus, it may be necessary to store both – perhaps a `<div>` with a special data attribute, with the understanding that it's autogenerated and overwritten from on a paired markdown source.

I'm not too familiar with `canvas-rce`, or TinyMCE itself, but a patch will get accepted, I'm willing to give implementation a shot, time-permitting.

**Comment by alalemp** [comment]
Bump on the request.
I have all instructions in markdown files (since I host them on Git)

**Comment by rdrpenguin04** [comment]
The idea linked was archived. I hope that someone from the open source community can move forward and attempt to implement this.
